### DISCUSSION TITLE
How to do integration testing between two applications (two different `TestClient`s)?

### DISCUSSION THREAD

**Discussion by UnoYakshi** [fastapi/fastapi#9236] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# app/settings.py
class Settings(BaseSettings):
    SQLALCHEMY_DATABASE_URL: str = ...
    FIRST_SERVICE_ADDRESS: str = 'http://localhost:8000'
    SECOND_SERVICE_ADDRESS: str = 'http://localhost:8001'


settings = Settings()  # type: ignore


# app/app.py
from fastapi import FastAPI
import requests


first_service = FastAPI()
second_service = FastAPI()


@first_service.get('/')
def ping_first() -> str:
    response = requests.get(settings.SECOND_SERVICE_ADDRESS)
    return f'pong1 + {response.content}'


@second_service.get('/')
def ping_second() -> str:
    return 'pong2'


# app/db.py
from sqlmodel import Session, create_engine

from app.settings import settings


engine = create_engine(settings.SQLALCHEMY_DATABASE_URL, echo=True)


def get_db_session() -> Session:
    with Session(engine) as session:
        yield session


# tests/conftest.py
import pytest
from sqlalchemy.orm import sessionmaker

from app.db import engine


TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)


def override_get_db():
    try:
        db = TestingSessionLocal()
        yield db
    finally:
        db.rollback()
        db.close()


@pytest.fixture(scope='session', autouse=True)
def first_service_client():
    from app.setting import settings

    first_service.dependency_overrides[get_db_session] = override_get_db

    with TestClient(
            app=first_service,
            base_url=settings.FIRST_SERVICE_ADDRESS,
            headers={"Content-Type": "application/json"},
    ) as client:
        yield client


@pytest.fixture(scope='session', autouse=True)
def second_service_client():
    from app.settings import settings

    second_service.dependency_overrides[get_db_session] = override_get_db

    with TestClient(
            app=second_service,
            base_url=settings.SECOND_SERVICE_ADDRESS,
            headers={"Content-Type": "application/json"},
    ) as client:
        yield client


# tests/smoke_test.py

def test_parsing(stub_worker, first_service_client, second_service_client):
    """It fails, obviously..."""

    response = first_service.post(url='/')
    assert response.content == 'pong1 + pong2'
```

### Description

The deal is quite simple. From the first service I want to call (via `requests` or `httpx`) another service. The problem is that neither of the services are actually set-up (you can't actually reach them, which is fine for unit-testing), however, they also can't interact with each other.

So my question is what are the common practices for integration testing when it comes to inter-serivce REST-interaction? Can I even do it w/o actually starting both services?

### Operating System

Linux

### Operating System Details

Ubuntu 20.04 LTS, 64-bit

Intel i7-10510U
16 GB RAM

### FastAPI Version

"0.92.0" (but "0.87.0" in `pyproject.toml`)

### Python Version

Python 3.8.7

### Additional Context

I understand this question probably should go to Starlette. But I presume FastAPI has bigger community.

**Comment by UnoYakshi** [comment]
So far, companies [I've talked to] usually just have the services running like usual in/on the testing environment/stage to do integration testing. And I'm not sure it's the best way possible.

**Comment by Kludex** [comment, marked as answer]
Yes: https://kludex.github.io/blog/2023/03/14/contract-testing-with-httpx-part-2/

**Reply by Kludex** [reply]
Feedback is welcome on that blog post btw :+1:

**Reply by UnoYakshi** [reply]
I'll definitely check it out! Thank you!

**Reply by UnoYakshi** [reply]
I'm not sure it will work in my real case as is. Right now, the call from Service A to Service B is buried underneath a few layers of code; I don't use a client `Dependency()` to call upon another service — it's just a simple `requests.post(...)`.

I'm sure a workaround is possible, but I'd prefer not to have another public route (from A to B). Also, there is a chance it won't be only my FastAPI services, so a simpler REST-call is preferable.
The workarounds I can imagine:
1) a private/protected route in Service A, i.e., it won't be accessible publicly, in `<service A>/docs`, etc.
2) if it's possible, have an overridable method for Serivce A that won't be a route; it looks like your `service_a.py::service_b_client()` is exactly that, though, — I guess I just don't entirely grasp how dependencies work. Because, aside the function is defined in the module file and being used as a `Dependency()`, I don't see how it's connected to the Service A or its router...

I'll try it out in my real (i.e., not simplified) code and will get back.

**Reply by UnoYakshi** [reply]
Aside from that, is it considered as a good and standard practice to call another of your services through a client (like in a code snap from your article/tutorial)? Wouldn't it be less bulky to use a REST-call w/o a client wrapping (`requests.get(...)`)?


```python3
async def service_b_client() -> AsyncIterator[httpx.AsyncClient]:
    async with httpx.AsyncClient(base_url="http://localhost:8000/b/") as client:
        yield client


@router.get("/call_b")
async def call_b(client: httpx.AsyncClient = Depends(service_b_client)):
    response = await client.get("/")
    return response.json()
```

**Reply by mdczaplicki** [reply]
The link's dead.

**Reply by jpvanegasc** [reply]
New link: https://fastapiexpert.com/blog/2023/03/14/contract-testing-with-httpx---part-2/
